samples/secure_tunneling/secure_tunnel/main.cpp

```diff
@@ -0,0 +1,278 @@
+                {
+                    fprintf(stdout, "Send Data Complete in Source Mode\n");
+                }
+                else
+                {
+                    fprintf(stderr, "Send Data Failed with error code %d\n", error_code);
+                }
+                break;
+        }
+    };
+
+    auto OnDataReceive = [&](const struct aws_byte_buf &data) {
+        receivedData = std::string((char *)data.buffer, data.len);
+        string returnMessage = "Echo:" + receivedData;
+
+        fprintf(stdout, "Received: \"%s\"\n", receivedData.c_str());
+
+        switch (localProxyMode)
+        {
+            case AWS_SECURE_TUNNELING_DESTINATION_MODE:
+                fprintf(stdout, "Data Receive Complete in Destination\n");
+                fprintf(stdout, "Sending response message:\"%s\"\n", returnMessage.c_str());
+                mSecureTunnel->SendData(ByteCursorFromCString(returnMessage.c_str()));
+                break;
+            case AWS_SECURE_TUNNELING_SOURCE_MODE:
+                fprintf(stdout, "Data Receive Complete in Source\n");
+                break;
+        }
+    };
+
+    /*
+     * This only fires in Destination Mode
+     */
+    auto OnStreamStart = [&]() { fprintf(stdout, "Stream Started in Destination Mode\n"); };
+
+    auto OnStreamReset = [&]() { fprintf(stdout, "Stream Reset\n"); };
+
+    auto OnSessionReset = [&]() { fprintf(stdout, "Session Reset\n"); };
+
+    /*********************** Secure Tunnel Setup ***************************/
+
+    /*
+     * Create a new SecureTunnel using the SecureTunnelBuilder
+     */
+    mSecureTunnel =
+        SecureTunnelBuilder(
+            Aws::Crt::g_allocator, bootstrap, SocketOptions(), accessToken.c_str(), localProxyMode, endpoint.c_str())
+            .WithRootCa(caFile.c_str())
+            .WithOnConnectionComplete(OnConnectionComplete)
+            .WithOnConnectionShutdown(OnConnectionShutdown)
+            .WithOnSendDataComplete(OnSendDataComplete)
+            .WithOnDataReceive(OnDataReceive)
+            .WithOnStreamStart(OnStreamStart)
+            .WithOnStreamReset(OnStreamReset)
+            .WithOnSessionReset(OnSessionReset)
+            .Build();
+
+    if (!mSecureTunnel)
+    {
+        fprintf(stderr, "Secure Tunnel Creation failed");
+        exit(-1);
+    }
+
+    mSecureTunnel->Connect();
+    int messageCount = 0;
+
+    while (1)
+    {
+        std::this_thread::sleep_for(3000ms);
+        if (localProxyMode == AWS_SECURE_TUNNELING_SOURCE_MODE)
+        {
+            messageCount++;
+            string toSend = to_string(messageCount) + ": " + message.c_str();
+
+            if (!mSecureTunnel->SendData(ByteCursorFromCString(toSend.c_str())))
+            {
+                fprintf(stdout, "Sending Message:\"%s\"\n", toSend.c_str());
+                if (messageCount >= 5)
+                {
+                    fprintf(stdout, "Closing Connection\n");
+                    mSecureTunnel->Close();
```

Comment:
like Connect(), the Close() function can also fail. probably want to print the error and exit if it fails

Reply:
Also, like Connect() the Close() call is async. Probably want to use another `std::promise` so that you wait for the connection closing to complete before exiting the program (currently, you're still looping every 3 seconds, it's just that the connection *tends to* finish before that.